src/emulator/tasksEmulator.ts

```diff
@@ -0,0 +1,314 @@
+    this.queues[key].delete(taskId);
+  }
+
+  createQueue(key: string, config: TaskQueueConfig): void {
+    const newQueue = new TaskQueue(key, config);
+    const intervalID = setInterval(
+      () => newQueue.refillTokens(),
+      TaskQueueController.TOKEN_REFRESH_INTERVAL,
+    );
+    this.tokenRefillIds.push(intervalID);
+    this.queues[key] = newQueue;
+  }
+
+  /**
+   * If there are no active queues (a queue is active if it has tasks in the queue or dispatch) then
+   * wait longer (1s) before checking the status of the queues again. If there are active queues,
+   * continuously (1ms) call their methods to handle dispatching tasks
+   */
+  listen(): void {
+    let shouldUpdate = false;
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    for (const [_key, queue] of Object.entries(this.queues)) {
+      shouldUpdate = shouldUpdate || queue.isActive();
+    }
+    if (shouldUpdate) {
+      this.updateQueues();
+      this.listenId = setTimeout(() => this.listen(), TaskQueueController.UPDATE_TIMEOUT);
+    } else {
+      this.listenId = setTimeout(() => this.listen(), TaskQueueController.LISTEN_TIMEOUT);
+    }
+  }
+
+  updateQueues(): void {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    for (const [_key, queue] of Object.entries(this.queues)) {
+      if (queue.isActive()) {
+        queue.dispatchTasks();
+        queue.processDispatch();
+      }
+    }
+  }
+
+  start(): void {
+    this.running = true;
+    this.listen();
+  }
+
+  stop(): void {
+    if (this.listenId) {
+      clearTimeout(this.listenId);
+    }
+    this.tokenRefillIds.forEach(clearInterval);
+    this.running = false;
+  }
+
+  isRunning(): boolean {
+    return this.running;
+  }
+
+  getStatistics() {
+    const stats: Record<string, any> = {};
+    for (const [key, queue] of Object.entries(this.queues)) {
+      stats[key] = queue.getStatistics();
+    }
+    return stats;
+  }
+}
+
+export class TasksEmulator implements EmulatorInstance {
+  private destroyServer?: () => Promise<void>;
+  private controller: TaskQueueController;
+
+  constructor(private args: TasksEmulatorArgs) {
+    this.controller = new TaskQueueController();
+  }
+
+  logger = EmulatorLogger.forEmulator(Emulators.TASKS);
+
+  createHubServer(): express.Application {
+    const hub = express();
+
+    const createTaskQueueRoute = `/projects/:project_id/locations/:location_id/queues/:queue_name`;
+    const createTaskQueueHandler: express.Handler = (req, res) => {
+      const projectId = req.params.project_id;
+      const locationId = req.params.location_id;
+      const queueName = req.params.queue_name;
+      const key = `queue:${projectId}-${locationId}-${queueName}`;
+      this.logger.logLabeled("SUCCESS", "tasks", `Created queue with key: ${key}`);
+      const body = req.body as TaskQueueConfig;
+      const taskQueueConfig: TaskQueueConfig = {
+        retryConfig: {
+          maxAttempts: body.retryConfig?.maxAttempts ?? 3,
```

Comment:
Sounds good. Also I got the defaults from the documentation in the functions SDK:
https://github.com/firebase/firebase-functions/blob/5145e188acfc73cbd195efa937a24e63c1a07246/src/common/providers/tasks.ts#L32

The Cloud Tasks documentation is a little vague just saying that it will "Set the default" https://cloud.google.com/tasks/docs/reference/rpc/google.cloud.tasks.v2#ratelimits

Or maybe it says the defaults somewhere else and I just cant find them.